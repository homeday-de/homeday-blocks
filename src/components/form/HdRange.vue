<template>
  <div
    :class="fieldClasses"
    class="range field"
  >
    <input type="range"
      v-model.number="computedValue"
      :id="name"
      :name="name"
      :required="required"
      :disabled="disabled"
      :min="min"
      :max="max"
      :step="step"
      @focus="focusHandler"
      @blur="blurHandler"
    >
    <div
      ref="track"
      :style="{
        background: trackBackground,
      }"
      class="range__track"
    >
      <div
        ref="progress"
        :style="{
          background: progressBackground,
        }"
        class="range__progress"
      />
    </div>
    <div v-if="displayStepBullets" class="range__steps">
      <button
        v-for="(_, stepIndex) in stepsAmount"
        :key="stepIndex"
        type="button"
        class="range__step"
        @click="onStepClick(stepIndex)"
      >
        <p
          v-if="labels[stepIndex]"
          class="range__step-label"
          v-text="labels[stepIndex]"
        />
      </button>
    </div>
    <div class="range__thumb" ref="thumb">
      <div
        v-if="displayTooltip"
        class="range__tooltip"
      >
        {{ tooltipValue || value }}
      </div>
      <div class="range__thumb__inner" ref="thumbInner">
        <div class="range__thumb__bullet" />
      </div>
    </div>
  </div>
</template>

<script>
import onResize from 'homeday-blocks/src/services/on-resize';

export default {
  name: 'HdRange',
  props: {
    name: {
      type: String,
      default: 'range',
    },
    required: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    min: {
      type: Number,
      default: 0,
    },
    max: {
      type: Number,
      default: 100,
    },
    step: {
      type: Number,
      default: 1,
    },
    value: {
      type: Number,
      default: 0,
    },
    labels: {
      type: Array,
      default: () => [],
    },
    displayStepBullets: {
      type: Boolean,
      default: false,
    },
    displayTooltip: {
      type: Boolean,
      default: false,
    },
    tooltipValue: {
      type: [String, Number],
      default: '',
    },
    trackBackground: {
      type: String,
      default: '', // falls back to the internal styles
    },
    progressBackground: {
      type: String,
      default: '', // falls back to the internal styles
    },
  },
  data() {
    return {
      isActive: null,
      trackWidth: 0,
    };
  },
  computed: {
    computedValue: {
      get() {
        return this.value;
      },
      set(value) {
        if (value !== this.computedValue) {
          this.$emit('input', value);
        }
      },
    },
    fieldClasses() {
      return {
        'field--active': this.isActive,
        'field--disabled': this.disabled,
        hasTooltip: this.displayTooltip,
      };
    },
    stepsAmount() {
      return Math.ceil(1 + (this.max - this.min) / this.step);
    },
  },
  watch: {
    min() {
      this.updateUI();
    },
    max() {
      this.updateUI();
    },
    step() {
      this.updateUI();
    },
    value() {
      this.adjustValue();
      this.updateUI();
    },
  },
  mounted() {
    this.adjustValue();

    this.trackWidth = this.$refs.track.offsetWidth;
    this.updateUI();
    onResize.onDebounced(this.onResize);
  },
  beforeDestroy() {
    onResize.offDebounced(this.onResize);
  },
  methods: {
    adjustValue() {
      let adjustedValue = this.value;

      adjustedValue -= adjustedValue % this.step;
      adjustedValue = Math.min(adjustedValue, this.max);
      adjustedValue = Math.max(adjustedValue, this.min);

      this.computedValue = adjustedValue;
    },
    updateUI() {
      // the percentage of the value, between max and min. I.e. : 15 is the 0.5 between 10 and 20
      const valuePercentage = (this.computedValue - this.min) / (this.max - this.min);
      const valuePercentageInputWidthPixels = this.trackWidth * valuePercentage;

      this.$refs.progress.style.transform = `scaleX(${valuePercentage})`;
      this.$refs.thumb.style.transform = `translateX(${valuePercentageInputWidthPixels}px)`;
    },
    onStepClick(stepIndex) {
      this.computedValue = this.min + (stepIndex * this.step);
    },
    onResize() {
      this.trackWidth = this.$refs.track.offsetWidth;
      this.updateUI();
    },
    focusHandler() {
      this.isActive = true;
    },
    blurHandler() {
      this.isActive = false;
    },
  },
};
</script>

<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.range {
  $range: &;
  height: $stack-l;
  position: relative;
  display: flex;
  align-items: center;

  &:focus {
    outline: none;
  }

  &.hasTooltip {
    margin-top: $stack-xl;
  }

  input[type=range] {
    -moz-appearance: none;
    -webkit-appearance: none;
    background: transparent;
    position: absolute;
    top: 0;
    left: 0;
    z-index: 2;
    width: 100%;
    margin: 0;
    padding: 0;
    height: $range-thumb-size;

    &:focus {
      outline: none;
    }

    // range invisible
    @mixin thumb-style {
      -webkit-appearance: none;
      background: none;
      border: none;
      height: $range-thumb-tot-size;
      width: $range-thumb-tot-size;
      display: block;
      -moz-appearance: none;
      cursor: grab;

      height: 1px;
      width: 1px;
      transform: scale(strip-unit($range-thumb-tot-size));
    }

    @mixin track-style {
      background: transparent;
      border-color: transparent;
      border: 0;
      color: transparent;
      cursor: pointer;

    }

    &::-webkit-slider-thumb {
      @include thumb-style;
    }

    &::-moz-range-thumb {
      @include thumb-style;
    }

    &::-ms-thumb {
      @include thumb-style;
    }

    &::-webkit-slider-runnable-track {
      @include track-style;
    }

    &::-moz-range-track {
      @include track-style;
    }

    &::-ms-track,
    &::-ms-fill-lower,
    &::-ms-fill-upper {
      @include track-style;
    }
  }

  &__track {
    position: absolute;
    top: 50%;
    margin-top: - 3px;
    width: 100%;
    height: $range-bar-height;
    border-radius: $range-bar-height;
    background: $disabledColor;
    z-index: 0;
    overflow: hidden;

    &:focus{
      box-shadow: none;
    }
  }

  &__steps {
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }

  &__step {
    height: 1px;
    width: 1px;
    padding: 0;
    background: transparent;
    border: 0;

    &:after {
      content: "";
      display: block;
      margin-left: - $steps-size / 2;
      margin-top: - $steps-size / 2;
      height: $steps-size - ($steps-border-size * 2);
      width: $steps-size - ($steps-border-size * 2);
      border-radius: $steps-size / 2;
      background: $primary-bg;
      border: $steps-border-size solid $quaternary-color;
      border-radius: 50%;
    }
  }

  &__step-label {
    position: absolute;
    top: #{$stack-l + $stack-s};
    transform: translateX(-50%);
    @include font('text-xsmall');
    color: getShade($quaternary-color, 80);
  }

  &__thumb {
    display: block;
    width: 0;
    height: 0;
    transition: transform .3s;

    &__inner {
      cursor: grab;
      box-shadow: $shadow;
      border: $range-thumb-border solid $activeColor;
      background: $disabledColor;
      height: $range-thumb-tot-size;
      width: $range-thumb-tot-size;
      border-radius: $range-thumb-size;
      position: absolute;
      background: $white;
      left: 0;
      margin-top: - $range-thumb-tot-size / 2;
      margin-left: - $range-thumb-tot-size / 2;
      z-index: 1;
      display: flex;
      align-items: center;
      justify-content: center;
    }

    .field--disabled & {
      border-color: $disabledColor;
      pointer-events: none;
    }

    &__bullet {
      background: $disabledColor;
      position: absolute;
      height: $range-thumb-size;
      width: $range-thumb-size;
      max-height: $range-thumb-inner-bullet-size;
      max-width: $range-thumb-inner-bullet-size;

      border-radius: $inset-m;
      transition: transform .1s ease-in-out;
      transform-origin: center center;


      #{$range}:not(.field--disabled):hover &{
        background-color: $activeColor;
        transform: scale(1.25);
      }

      @include only-ie {
        top: 50%;
        left: 50%;
        margin-left: - $range-thumb-inner-bullet-size / 2;
        margin-top: - $range-thumb-inner-bullet-size / 2;
      }
    }
  }

  &__progress {
    background-color: $activeColor;
    height: 100%;
    width: 100%;
    transform-origin: 0 center;
    transform: scaleX(0);
    transition: transform .3s;

    .field--disabled & {
      background-color: $disabledColor;
    }
  }

  &__tooltip {
    position: absolute;
    bottom: $stack-m;
    transform: translateX(-50%);
    color: $white;
    background: url('~homeday-blocks/src/assets/tooltip.svg') no-repeat;
    background-size: 100% 100%;
    padding: $stack-s $inline-m #{$stack-m + $stack-s};
    pointer-events: none;
  }
}
</style>
