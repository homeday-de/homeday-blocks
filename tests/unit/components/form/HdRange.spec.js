import {
  wrapperFactoryBuilder,
  getLastEventPayload,
} from 'tests/unit/helpers';
import HdRange from '@/components/form/HdRange.vue';

const TRACK_SELECTOR = '.range__track';
const PROGRESS_SELECTOR = '.range__progress';
const STEP_SELECTOR = '.range__step';
const TOOLTIP_SELECTOR = '.range__tooltip';
const LABEL_SELECTOR = '.range__step-label';

const initialValue = 0;
const max = 100;
const min = -100;
const step = 50;

const wrapperBuilder = wrapperFactoryBuilder(HdRange, {
  props: {
    name: 'testRange',
    min,
    max,
    step,
    value: initialValue,
    displayStepBullets: true,
  },
});

describe('HdRange', () => {
  it('renders as expected', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders component with preselected value', () => {
    const wrapper = wrapperBuilder();

    expect(parseFloat(wrapper.find('input').element.value)).toBe(initialValue);
  });

  it('limits the value to the passed `min` and `max`', async () => {
    const wrapper = wrapperBuilder({
      props: {
        value: max + 1,
      },
    });

    expect(getLastEventPayload({ wrapper, eventName: 'input' })).toBe(max);

    wrapper.setProps({
      value: min - 1,
    });

    await wrapper.vm.$nextTick();

    expect(getLastEventPayload({ wrapper, eventName: 'input' })).toBe(min);
  });

  it('on input blur, the proper status is set and the proper method is executed', () => {
    const wrapper = wrapperBuilder();

    wrapper.get('input').trigger('focus');

    expect(wrapper.vm.isActive).toBe(true);
  });

  it('on input blur, the proper status is set and the proper method is executed', () => {
    const wrapper = wrapperBuilder({
      data: {
        isActive: true,
      },
    });

    wrapper.get('input').trigger('blur');

    expect(wrapper.vm.isActive).toBe(false);
  });

  it('on props change, UI is updated', async () => {
    const mockUpdateUI = jest.fn();
    const wrapper = wrapperBuilder({
      props: {
        min: min + 1,
      },
      methods: {
        updateUI: mockUpdateUI,
      },
    });

    expect(mockUpdateUI).toHaveBeenCalledTimes(1);

    wrapper.setProps({
      max: max - 1,
    });

    await wrapper.vm.$nextTick();

    expect(mockUpdateUI).toHaveBeenCalledTimes(2);

    wrapper.setProps({
      step: 10,
    });

    await wrapper.vm.$nextTick();

    expect(mockUpdateUI).toHaveBeenCalledTimes(3);
  });

  it('step bullets are properly displayed', async () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.findAll(STEP_SELECTOR).length).toBe(5);

    wrapper.setProps({
      displayStepBullets: false,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.findAll(STEP_SELECTOR).length).toBe(0);
  });

  it('emits the right value on steps click', () => {
    const wrapper = wrapperBuilder();
    const steps = wrapper.findAll(STEP_SELECTOR);

    steps.at(0).trigger('click');
    expect(getLastEventPayload({ wrapper, eventName: 'input' })).toBe(min);

    steps.at(steps.length - 1).trigger('click');
    expect(getLastEventPayload({ wrapper, eventName: 'input' })).toBe(max);
  });

  describe('Labels', () => {
    const TEST_LABELS = ['label1', 'label<br>2', 'label3', 'label4'];
    const TEST_INDEX = Math.round(TEST_LABELS.length / 2);
    const labelsWrapperBuilder = wrapperFactoryBuilder(HdRange, {
      props: {
        name: 'testrange',
        min,
        max,
        step,
        value: initialValue,
        displayStepBullets: true,
        labels: TEST_LABELS,
      },
    });

    it('renders', () => {
      const wrapper = labelsWrapperBuilder();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders the right number of labels', () => {
      const wrapper = labelsWrapperBuilder();

      expect(wrapper.findAll(LABEL_SELECTOR).length).toBe(TEST_LABELS.length);
    });

    it('shows the right values', () => {
      const wrapper = labelsWrapperBuilder();

      expect(wrapper.findAll(LABEL_SELECTOR).at(TEST_INDEX).text()).toBe(TEST_LABELS[TEST_INDEX]);
    });
  });

  describe('Tooltip', () => {
    const tooltipWrapperBuilder = wrapperFactoryBuilder(HdRange, {
      props: {
        name: 'testrange',
        min,
        max,
        step,
        value: initialValue,
        displayStepBullets: true,
        displayTooltip: true,
      },
    });

    it('is visible when the `displayTooltip` prop is set to `true`', async () => {
      const wrapper = tooltipWrapperBuilder();

      expect(wrapper.find(TOOLTIP_SELECTOR).exists()).toBe(true);

      wrapper.setProps({
        displayTooltip: false,
      });

      await wrapper.vm.$nextTick();

      expect(wrapper.find(TOOLTIP_SELECTOR).exists()).toBe(false);
    });

    it('accepts custom value', () => {
      const TEST_TOOLTIP_VALUE = 'Foo Bar';
      const wrapper = tooltipWrapperBuilder({
        props: {
          tooltipValue: TEST_TOOLTIP_VALUE,
        },
      });

      expect(wrapper.find(TOOLTIP_SELECTOR).text()).toBe(TEST_TOOLTIP_VALUE);
    });

    it('shows the range value as a fallback', () => {
      const TEST_VALUE = 10;
      const wrapper = tooltipWrapperBuilder({
        props: {
          displayTooltip: true,
          value: TEST_VALUE,
        },
      });

      expect(wrapper.find(TOOLTIP_SELECTOR).text()).toBe(TEST_VALUE.toString());
    });
  });

  it('supports setting custom backgrounds', () => {
    const TEST_COLOR1 = 'red';
    const TEST_COLOR2 = 'blue';
    const wrapper = wrapperBuilder({
      props: {
        trackBackground: TEST_COLOR1,
        progressBackground: TEST_COLOR2,
      },
    });

    expect(wrapper.find(TRACK_SELECTOR).attributes().style).toMatch(`background: ${TEST_COLOR1}`);
    expect(wrapper.find(PROGRESS_SELECTOR).attributes().style).toMatch(`background: ${TEST_COLOR2}`);
  });

  it('calls `updateUI` on resize', () => {
    const mockedUpdateUI = jest.fn();
    const wrapper = wrapperBuilder({
      methods: {
        updateUI: mockedUpdateUI,
      },
    });

    wrapper.vm.onResize();

    expect(mockedUpdateUI).toHaveBeenCalledTimes(2);
  });
});
