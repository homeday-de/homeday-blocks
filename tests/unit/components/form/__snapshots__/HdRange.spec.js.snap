// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdRange renders component with preselected value 1`] = `
<div class="range field"><input type="range" id="testRange" name="testRange" min="-100" max="100" step="1">
  <div class="range__decoration">
    <div class="range__progress" style="width: 1px; transform: scaleX(0);"></div>
  </div>
  <ul class="range__steps" style="padding-right: 0%;">
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
    <li></li>
  </ul>
  <div class="range__thumb" style="transform: translateX(0px);">
    <div class="range__thumb__inner">
      <div class="range__thumb__bullet"></div>
    </div>
  </div>
</div>
`;
